File: fixtures/scale/dart/large/lib/ciphers/simon/simon_rounds.dart
import 'simon_key.dart';

/// Simon32-style Feistel rounds built from AND, rotate and xor.
int encryptBlock(int block, SimonKey key) {
  var v0 = (block >> 32) & 0xFFFFFFFF;
  var v1 = block & 0xFFFFFFFF;
  for (var r = 0; r < 32; r++) {
    final tmp = v0;
    final rot1 = (v0 << 1 | v0 >> 31) & 0xFFFFFFFF;
    final rot8 = (v0 << 8 | v0 >> 24) & 0xFFFFFFFF;
    final rot2 = (v0 << 2 | v0 >> 30) & 0xFFFFFFFF;
    v0 = (v1 ^ (rot1 & rot8) ^ rot2 ^ key.k(r & 3)) & 0xFFFFFFFF;
    v1 = tmp;
  }
  return _pack(v0, v1);
}

int decryptBlock(int block, SimonKey key) {
  var v0 = (block >> 32) & 0xFFFFFFFF;
  var v1 = block & 0xFFFFFFFF;
  for (var r = 32 - 1; r >= 0; r--) {
    final tmp = v1;
    final rot1 = (v1 << 1 | v1 >> 31) & 0xFFFFFFFF;
    final rot8 = (v1 << 8 | v1 >> 24) & 0xFFFFFFFF;
    final rot2 = (v1 << 2 | v1 >> 30) & 0xFFFFFFFF;
    v1 = (v0 ^ (rot1 & rot8) ^ rot2 ^ key.k(r & 3)) & 0xFFFFFFFF;
    v0 = tmp;
  }
  return _pack(v0, v1);
}

int _pack(int v0, int v1) => ((v0 & 0xFFFFFFFF) << 32) | (v1 & 0xFFFFFFFF);

File: fixtures/scale/scala/large/src/cryptobench/ciphers/simon/SimonRounds.scala
package cryptobench.ciphers.simon

/** Simon64-style Feistel rounds built from AND, rotate and xor. */
private[simon] object SimonRounds {
  def encryptBlock(block: Long, key: SimonKey): Long = {
    var v0 = (block >>> 32).toInt
    var v1 = block.toInt
    for (r <- 0 until 32) {
      val tmp = v0
      v0 = v1 ^ (Integer.rotateLeft(v0, 1) & Integer.rotateLeft(v0, 8)) ^ Integer.rotateLeft(v0, 2) ^ key.k(r & 3)
      v1 = tmp
    }
    pack(v0, v1)
  }

  def decryptBlock(block: Long, key: SimonKey): Long = {
    var v0 = (block >>> 32).toInt
    var v1 = block.toInt
    for (r <- 32 - 1 to 0 by -1) {
      val tmp = v1
      v1 = v0 ^ (Integer.rotateLeft(v1, 1) & Integer.rotateLeft(v1, 8)) ^ Integer.rotateLeft(v1, 2) ^ key.k(r & 3)
      v0 = tmp
    }
    pack(v0, v1)
  }

  private def pack(v0: Int, v1: Int): Long =
    (v0.toLong << 32) | (v1.toLong & 0xFFFFFFFFL)
}
